### DISCUSSION TITLE
Recommended way of integrating client-side hashing for signing PDFs

### DISCUSSION THREAD

**Discussion by ralienpp** [Keyfactor/signserver-ce#117] [General]
I'd like to use the client-side hashing feature for signing PDFs (currently this is only achievable by sending the entire file to the server).

What would be the criteria for integrating a pull request that provides this functionality? Specifically, what kind of external dependencies would be acceptable?

For example, there is [this library that has signing](https://pdfbox.apache.org/) functionality, which could hypothetically be used to do the client-side part of forming the data to be signed, and then taking care of embedding the signature into the PDF. While this could make signclient more versatile, it would also begin to turn it into a "kitchen sink".

What would be the recommended way of adding this functionality, but without having to invent a PDF parser from scratch? Is it possible to rely on this external dependency as an optional feature?

**Comment by netmackan** [comment]
As PDF parsing and signing is already supported on the server-side, the preferred option would be to use the same library also in SignClient.
